File: Assets/Scripts/DouyinSequenceOrbitUnit.lua
---@var OrbitCenter            :UnityEngine.Transform
---@var OrbitAxis              :Vector3
---@var OrbitLoops             :float = 2
---@var AccelDuration          :float = 1
---@var CruiseDuration         :float = 1
---@var DecelDuration          :float = 1
---@var KeepOriginalRotation   :bool = false
---@end

local baseWorldPosition = nil
local baseWorldRotation = nil
local cachedCenterPosition = nil
local currentAngle = 0
local elapsedTime = 0
local totalAngle = 0
local totalDuration = 0
local maxAngularSpeed = 0
local accelRate = 0
local decelRate = 0
local isPlaying = false
local isComplete = true
local normalizedAxis = nil
local hasInitialized = false

local function ClampNonNegative(value, defaultValue)
    if value == nil then
        return defaultValue
    end
    if value < 0 then
        return 0
    end
    return value
end

local function Clamp01(value)
    if value == nil then
        return 0
    end
    if value < 0 then
        return 0
    end
    if value > 1 then
        return 1
    end
    return value
end

local function NormalizeAxis(axis, fallbackAxis)
    local source = axis
    if source == nil or source.sqrMagnitude <= 0.0001 then
        source = fallbackAxis
    end

    local magnitude = math.sqrt(source.x * source.x + source.y * source.y + source.z * source.z)
    if magnitude <= 0.0001 then
        return fallbackAxis
    end

    return CS.UnityEngine.Vector3(source.x / magnitude, source.y / magnitude, source.z / magnitude)
end

local function RefreshAxis()
    normalizedAxis = NormalizeAxis(OrbitAxis, CS.UnityEngine.Vector3.up)
end

local function RebuildMotionProfile()
    OrbitLoops = ClampNonNegative(OrbitLoops, 2)
    AccelDuration = ClampNonNegative(AccelDuration, 1)
    CruiseDuration = ClampNonNegative(CruiseDuration, 1)
    DecelDuration = ClampNonNegative(DecelDuration, 1)

    totalAngle = 360 * OrbitLoops
    totalDuration = AccelDuration + CruiseDuration + DecelDuration

    local effectiveTime = 0.5 * AccelDuration + CruiseDuration + 0.5 * DecelDuration
    if effectiveTime <= 0.0001 or totalAngle <= 0.0001 then
        maxAngularSpeed = 0
        accelRate = 0
        decelRate = 0
        return
    end

    maxAngularSpeed = totalAngle / effectiveTime
    accelRate = AccelDuration > 0 and (maxAngularSpeed / AccelDuration) or 0
    decelRate = DecelDuration > 0 and (maxAngularSpeed / DecelDuration) or 0
end

local function EnsureInitialized()
    if KeepOriginalRotation == nil then
        KeepOriginalRotation = false
    end
    if normalizedAxis == nil then
        RefreshAxis()
    end
    RebuildMotionProfile()
    if baseWorldPosition == nil then
        baseWorldPosition = self.transform.position
    end
    if baseWorldRotation == nil then
        baseWorldRotation = self.transform.rotation
    end
    hasInitialized = true
end

local function CaptureBaseState()
    baseWorldPosition = self.transform.position
    baseWorldRotation = self.transform.rotation
end

local function ResolveCenterPosition()
    if OrbitCenter ~= nil then
        return OrbitCenter.position
    end
    return baseWorldPosition
end

local function GetAngleAtTime(time)
    if time <= 0 then
        return 0
    end

    if totalDuration <= 0.0001 or totalAngle <= 0.0001 then
        return totalAngle
    end

    local angle = 0
    local t = time

    if AccelDuration > 0 then
        if t < AccelDuration then
            return 0.5 * accelRate * t * t
        end
        angle = angle + 0.5 * maxAngularSpeed * AccelDuration
        t = t - AccelDuration
    end

    if CruiseDuration > 0 then
        if t < CruiseDuration then
            return angle + maxAngularSpeed * t
        end
        angle = angle + maxAngularSpeed * CruiseDuration
        t = t - CruiseDuration
    end

    if DecelDuration > 0 and t < DecelDuration then
        return angle + maxAngularSpeed * t - 0.5 * decelRate * t * t
    end

    return totalAngle
end

local function ApplyAngle(targetAngle)
    local deltaAngle = targetAngle - currentAngle
    if math.abs(deltaAngle) <= 0.0001 then
        return
    end

    self.transform:RotateAround(cachedCenterPosition, normalizedAxis, deltaAngle)
    if KeepOriginalRotation then
        self.transform.rotation = baseWorldRotation
    end
    currentAngle = targetAngle
end

function Awake()
    local alreadyInitialized = hasInitialized
    EnsureInitialized()
    if not alreadyInitialized then
        CaptureBaseState()
        ResetToBase()
    end
end

function ResetToBase()
    EnsureInitialized()

    self.transform.position = baseWorldPosition
    self.transform.rotation = baseWorldRotation
    cachedCenterPosition = ResolveCenterPosition()
    elapsedTime = 0
    currentAngle = 0
    isPlaying = false
    isComplete = true
end

function BeginForward()
    EnsureInitialized()
    ResetToBase()
    isComplete = false

    if totalDuration <= 0.0001 or totalAngle <= 0.0001 then
        isPlaying = false
        isComplete = true
        return
    end

    isPlaying = true
end

function BeginBackward()
    BeginForward()
end

function Tick(deltaTime)
    EnsureInitialized()
    if not isPlaying or isComplete then
        return
    end

    elapsedTime = elapsedTime + deltaTime
    if elapsedTime >= totalDuration then
        self.transform.position = baseWorldPosition
        self.transform.rotation = baseWorldRotation
        elapsedTime = totalDuration
        currentAngle = 0
        isPlaying = false
        isComplete = true
        return
    end

    ApplyAngle(GetAngleAtTime(elapsedTime))
end

function IsComplete()
    return isComplete
end

function GetProgress01()
    if totalDuration <= 0.0001 then
        return 1
    end
    if isComplete then
        return 1
    end
    return Clamp01(elapsedTime / totalDuration)
end

function GetTotalDuration()
    return totalDuration
end

function GetSpeedFactor()
    if maxAngularSpeed <= 0.0001 or totalDuration <= 0.0001 then
        return 0
    end

    local t = elapsedTime
    if t <= 0 then
        return 0
    end

    if AccelDuration > 0 then
        if t < AccelDuration then
            return Clamp01(t / AccelDuration)
        end
        t = t - AccelDuration
    end

    if CruiseDuration > 0 then
        if t < CruiseDuration then
            return 1
        end
        t = t - CruiseDuration
    end

    if DecelDuration > 0 and t < DecelDuration then
        return Clamp01(1 - (t / DecelDuration))
    end

    return 0
end

File: Assets/Scripts/DouyinSequenceOrbitUnit.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using UnityEngine;

public sealed class DouyinSequenceOrbitUnit : MonoBehaviour, IDouyinSequenceUnit
{
    public Transform OrbitCenter;
    public Vector3 OrbitAxis = Vector3.up;
    public float OrbitLoops = 2f;
    public float AccelDuration = 1f;
    public float CruiseDuration = 1f;
    public float DecelDuration = 1f;
    public bool KeepOriginalRotation;

    private Vector3 baseWorldPosition;
    private Quaternion baseWorldRotation;
    private Vector3 cachedCenterPosition;
    private float currentAngle;
    private float elapsedTime;
    private float totalAngle;
    private float totalDuration;
    private float maxAngularSpeed;
    private float accelRate;
    private float decelRate;
    private bool isPlaying;
    private bool isComplete = true;
    private Vector3 normalizedAxis;
    private bool hasInitialized;

    private void Awake()
    {
        bool alreadyInitialized = hasInitialized;
        EnsureInitialized();
        if (!alreadyInitialized)
        {
            CaptureBaseState();
            ResetToBase();
        }
    }

    public void ResetToBase()
    {
        EnsureInitialized();
        transform.position = baseWorldPosition;
        transform.rotation = baseWorldRotation;
        cachedCenterPosition = ResolveCenterPosition();
        elapsedTime = 0f;
        currentAngle = 0f;
        isPlaying = false;
        isComplete = true;
    }

    public void BeginForward()
    {
        EnsureInitialized();
        ResetToBase();
        isComplete = false;

        if (totalDuration <= 0.0001f || totalAngle <= 0.0001f)
        {
            isPlaying = false;
            isComplete = true;
            return;
        }

        isPlaying = true;
    }

    public void BeginBackward()
    {
        BeginForward();
    }

    public void Tick(float deltaTime)
    {
        EnsureInitialized();
        if (!isPlaying || isComplete)
        {
            return;
        }

        elapsedTime += deltaTime;
        if (elapsedTime >= totalDuration)
        {
            transform.position = baseWorldPosition;
            transform.rotation = baseWorldRotation;
            elapsedTime = totalDuration;
            currentAngle = 0f;
            isPlaying = false;
            isComplete = true;
            return;
        }

        ApplyAngle(GetAngleAtTime(elapsedTime));
    }

    public bool IsComplete() => isComplete;

    public float GetProgress01()
    {
        if (totalDuration <= 0.0001f || isComplete)
        {
            return 1f;
        }
        return Mathf.Clamp01(elapsedTime / totalDuration);
    }

    public float GetTotalDuration() => totalDuration;

    public float GetSpeedFactor()
    {
        if (maxAngularSpeed <= 0.0001f || totalDuration <= 0.0001f)
        {
            return 0f;
        }

        float t = elapsedTime;
        if (t <= 0f)
        {
            return 0f;
        }

        if (AccelDuration > 0f)
        {
            if (t < AccelDuration)
            {
                return Mathf.Clamp01(t / AccelDuration);
            }
            t -= AccelDuration;
        }

        if (CruiseDuration > 0f)
        {
            if (t < CruiseDuration)
            {
                return 1f;
            }
            t -= CruiseDuration;
        }

        if (DecelDuration > 0f && t < DecelDuration)
        {
            return Mathf.Clamp01(1f - t / DecelDuration);
        }

        return 0f;
    }

    private void EnsureInitialized()
    {
        normalizedAxis = NormalizeAxis(OrbitAxis, Vector3.up);
        RebuildMotionProfile();
        if (!hasInitialized)
        {
            baseWorldPosition = transform.position;
            baseWorldRotation = transform.rotation;
        }
        hasInitialized = true;
    }

    private void CaptureBaseState()
    {
        baseWorldPosition = transform.position;
        baseWorldRotation = transform.rotation;
    }

    private void RebuildMotionProfile()
    {
        OrbitLoops = Mathf.Max(0f, OrbitLoops);
        AccelDuration = Mathf.Max(0f, AccelDuration);
        CruiseDuration = Mathf.Max(0f, CruiseDuration);
        DecelDuration = Mathf.Max(0f, DecelDuration);

        totalAngle = 360f * OrbitLoops;
        totalDuration = AccelDuration + CruiseDuration + DecelDuration;

        float effectiveTime = 0.5f * AccelDuration + CruiseDuration + 0.5f * DecelDuration;
        if (effectiveTime <= 0.0001f || totalAngle <= 0.0001f)
        {
            maxAngularSpeed = 0f;
            accelRate = 0f;
            decelRate = 0f;
            return;
        }

        maxAngularSpeed = totalAngle / effectiveTime;
        accelRate = AccelDuration > 0f ? maxAngularSpeed / AccelDuration : 0f;
        decelRate = DecelDuration > 0f ? maxAngularSpeed / DecelDuration : 0f;
    }

    private Vector3 ResolveCenterPosition()
    {
        return OrbitCenter != null ? OrbitCenter.position : baseWorldPosition;
    }

    private float GetAngleAtTime(float time)
    {
        if (time <= 0f)
        {
            return 0f;
        }

        if (totalDuration <= 0.0001f || totalAngle <= 0.0001f)
        {
            return totalAngle;
        }

        float angle = 0f;
        float t = time;

        if (AccelDuration > 0f)
        {
            if (t < AccelDuration)
            {
                return 0.5f * accelRate * t * t;
            }
            angle += 0.5f * maxAngularSpeed * AccelDuration;
            t -= AccelDuration;
        }

        if (CruiseDuration > 0f)
        {
            if (t < CruiseDuration)
            {
                return angle + maxAngularSpeed * t;
            }
            angle += maxAngularSpeed * CruiseDuration;
            t -= CruiseDuration;
        }

        if (DecelDuration > 0f && t < DecelDuration)
        {
            return angle + maxAngularSpeed * t - 0.5f * decelRate * t * t;
        }

        return totalAngle;
    }

    private void ApplyAngle(float targetAngle)
    {
        float deltaAngle = targetAngle - currentAngle;
        if (Mathf.Abs(deltaAngle) <= 0.0001f)
        {
            return;
        }

        transform.RotateAround(cachedCenterPosition, normalizedAxis, deltaAngle);
        if (KeepOriginalRotation)
        {
            transform.rotation = baseWorldRotation;
        }
        currentAngle = targetAngle;
    }

    private static Vector3 NormalizeAxis(Vector3 axis, Vector3 fallbackAxis)
    {
        return axis.sqrMagnitude > 0.0001f ? axis.normalized : fallbackAxis;
    }
}

